src/app/data-model/DecodeWithoutFabricIndex.h

```diff
@@ -0,0 +1,66 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/data-model/DecodableList.h>
+#include <app/data-model/Decode.h>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/*
+ * @brief
+ *
+ * This function works just like Decode(reader, x), however, for fabric scoped data, decode with this won't overwrite the fabric
+ * index field in the payload. This function is expected to be called by client side code when decoding the values from server with
+ * a fabric index field with a unified API for both fabric-scoped and non-fabric-scoped values.
+ *
+ */
+template <typename X,
+          typename std::enable_if_t<std::is_class<X>::value &&
+                                        std::is_same<decltype(Decode(std::declval<TLV::TLVReader &>(), std::declval<X &>(),
+                                                                     std::declval<Optional<FabricIndex> &>())),
+                                                     CHIP_ERROR>::value,
+                                    X> * = nullptr>
+CHIP_ERROR DecodeWithoutFabricIndex(TLV::TLVReader & reader, X & x)
```

Comment:
I'd very much like to avoid this duality of Decode functions just to tip toe around over-writing the fabric index issue.

Why not just have a single Decode signature as before, that would attempt to decode the fabric index if present in the payload but it not present, would keep the field value to its default.

There-after, you can call a method on it to over-write/set the fabric index in the specific code paths that necessitate that.

Reply:
The users should be aware when not overwritting the fabric index field.
For example, for ACL clusters, if we called Decode without a fabric index, the client might be able to write an entry to a fabric it does not own. Similar to some fabric senstive values.

However, on read side, we will never override such field, this function will provide a unique API for both fabric scoped and non-fabric scoped structs.

Reply:
> if we called Decode without a fabric index, the client might be able to write an entry to a fabric it does not own

But server logic isn't _expected_ to directly call this `Decode` right? They would go through `AttributeValueDecoder::Decode`, which would correctly stuff in the fabric index **before** handing it to cluster logic.

Reply:
Based on this idea, I think it is ok ot provide two signatures of Decode() and remove the `DecodeWithoutFabricIndex`